python-api/livy/client.py

```diff
@@ -0,0 +1,444 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from ConfigParser import SafeConfigParser
+from StringIO import StringIO
+import os
+from concurrent.futures import ThreadPoolExecutor
+import json
+import cloudpickle
+import base64
+from job_handle import JobHandle
+import re
+from urlparse import ParseResult
+import traceback
+import requests
+import threading
+
+
+class HttpClient(object):
+    """A http based client for submitting Spark-based jobs to a Livy backend.
+
+    Parameters
+    ----------
+    uri : urlparse.ParseResult instance
+        Livy server uri generated from urlparse lib
+    load_defaults : boolean, optional
+        This parameter decides if the default config needs to be loaded
+        Default is True
+    conf_dict : dict, optional
+        The key-value pairs in the conf_dict will be loaded to the config
+
+    Examples
+    --------
+    Imports needed to create an instance of HttpClient
+    >>> from livy.client import HttpClient
+    >>> from urlparse import urlparse
+
+    1) Creates a client that is loaded with default config
+       as 'load_defaults' is True by default
+    >>> client = HttpClient(urlparse("http://example:8998/"))
+
+    2) Creates a client that does not load default config, but loads
+       config that are passed in 'config_dict'
+    >>> config_dict = {'spark.app.name', 'Test App'}
+    >>> client = HttpClient(urlparse("http://example:8998/"),
+    >>>             load_defaults=False, config_dict=config_dict)
+
+    """
+
+    _CONFIG_SECTION = 'env'
+    _LIVY_CLIENT_CONF_DIR = "LIVY_CLIENT_CONF_DIR"
+
+    def __init__(self, uri, **kwargs):
+        self._config = SafeConfigParser()
+        self._load_config(**kwargs)
+        match = re.match(r'(.*)/sessions/([0-9]+)', uri.path)
+        if match:
+            base = ParseResult(scheme=uri.scheme, netloc=uri.netloc,
+                               path=match.group(1), params=uri.params,
+                               query=uri.query, fragment=uri.fragment)
+            self._set_uri(base)
+            self._conn = LivyConnection(base)
+            self._session_id = int(match.group(2))
+            self._reconnect_to_existing_session()
+        else:
+            self._set_uri(uri)
+            session_conf_dict = dict(self._config.items(self._CONFIG_SECTION))
+            self._conn = LivyConnection(uri)
+            self._session_id = self._create_new_session(
+                session_conf_dict).json()['id']
+        self._executor = ThreadPoolExecutor(max_workers=1)
+        self._stopped = False
+
+    def submit(self, job):
+        """
+        Submits a job for execution to the spark cluster.
+
+        Parameters
+        ----------
+        job : function
+            The function must accept a single parameter, which is an instance
+            of JobContext. Spark jobs can be created with the help of
```

Comment:
This wording is confusing. `JobContext` is already documented, isn't it? So you could remove the second sentence here and avoid the confusion.